TheEthicalHackerCup/Assets/Scripts/Learning/QuizUIManager.cs

```diff
@@ -0,0 +1,172 @@
+            {
+                var x = new int();
+                x = i;
+                buttons[i].onClick.AddListener(() => handleOptionSelected(x));
+            }
+            nextButton.onClick.AddListener(handleNextSelected);
+
+
+
+        }
+
+        // Update is called once per frame
+        void Update()
+        {
+
+        }
+
+        private void showButtons(bool show)
+        {
+            int i = 0;
+            foreach (var button in buttons)
+            {
+                button.enabled = show;
+                var color = button.colors;
+                color.normalColor = Color.white;
+                color.selectedColor = Color.white;
+                color.highlightedColor = Color.white;
+                color.disabledColor = Color.white;
+                color.pressedColor = Color.white;
+                buttons[i].colors = color;
+                if (i++ != 0)
+                    button.gameObject.SetActive(show);
+            }
+        }
+
+        private void handleClosed(object sender, QuizClosedEvent evt)
+        {
+            Debug.Log(evt.pass ? "Quiz Passed" : "Quiz Failed");
+            SceneManager.LoadScene(mainScene);
+        }
+
+        private void handleStart(object sender, QuizStartedEvent evt)
+        {
+            title.text = evt.quizName;
+            showButtons(false);
+            buttons[0].GetComponentInChildren<TMP_Text>().text = "Welcome to the Learning game for " + evt.quizName + ". Press next (>) to start.";
+        }
+        private void handleNextQuestion(object sender, NextQuestionEvent evt)
+        {
+            this.current = evt.question;
+            this.current.QuestionStateUpdated += handleQuestionStateChanged;
+            this.current.invokeQuestionEvent(new StartEvent());
+        }
+        private void handleSubmitted(object sender, QuizSubmittedEvent evt)
+        {
+            showButtons(false);
+            title.text = "SCORE!!!";
+            var output = "Your score: " + string.Format("{0:F1}", evt.percentScore) + "%\n" +
+                "You " + (evt.pass ? "Passed!!" : " should review the material and try again next time :D\n" +
+                "Click next to continue back to the main game");
+            buttons[0].GetComponentInChildren<TMP_Text>().text = output;
+        }
+        private void handleQuestionStateChanged(object sender, QuestionStateUpdatedEvent evt)
+        {
+            showButtons(true);
+            if (evt.QuestionState.GetType() == typeof(CheckboxState))
+            {
+                this.handleCheckboxQuestion(sender, evt);
+            }
+            else if (evt.QuestionState.GetType() == typeof(RadioState))
+            {
+                this.handleRadioQuestion(sender, evt);
+            }
+            else
+            {
+                throw new System.Exception("Invalid State type");
+            }
+        }
+
+
+        private void handleCheckboxQuestion(object sender, QuestionStateUpdatedEvent evt)
```

Comment:
Avoid using object as a type for the sender param

This seems to be an event handler type of thing. Some ideas...
- Convert this to CSharpy event Action ways to send events around. You can make a whole lotta input params to ID the sender and you would not need so many event classes
OR
- Make an interface that is implemented by all classes that can be the sender param.  This helps the code be a lil more defensive because  object as a type is a lil bit of a free for all

Reply:
This is how EventHandlers work in C#